### PR TITLE
Added Union Segments command line

### PR DESCRIPTION
**Pull request by LeeTL1220**
Added command line to do a segment union.  Inputs:
- exactly two segment files (arbitrary column headers).
- columns of interest

What it does:
- Attempts to find reasonable headers in the TSVs
- Creates instances of a simplistic object that is composed of a mapping for columns and an interval.  Note that the columns making up the interval are not in the map.
- Unions the segments and the columns of interest into a new TSV.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 1]
```diff
@@ -1104,9 +1111,199 @@ static public int beginOfShard(int shardIndex, int shardSize) {
...
+        return result;
+    }
+
+    /**
+     * Creates a map of which locatables (keys) overlap the other list of locatables (vals)
+     *
+     * Input lists are assumed sorted by interval.  These are lists solely because we need the ordering to make
+     *  this run efficiently.
+     *
+     *  Within a single input list, segments cannot overlap (or duplicate).  If this occurs, behavior is undefined.
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: Again, better to validate and fail.  Especially since I do not think you check consistent sorting with respect to contigs in calling code.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            }
+
+            // Create a map of input to output headers.  I.e. the annotation in the segment1 to the output that should be written in the final file.
+            //  This assumes that the keys in each entry of the list is the same.
+            // TODO: If we want to support more than two segment files, this is the only bit that requires thinking.  Once this is solved, then this logic can go into a utility class.
+            final Set<String> intersectingAnnotations = Sets.intersection(segments1.get(0).getAnnotations().keySet(), segments2.get(0).getAnnotations().keySet());
+
+            // Create the obvious mappings that are identity then tack on new annotations for conflicts.
+            // These are mappings that take the header name from the segment files and map to an output header to avoid conflicts.
+            final Map<String, String> input1ToOutputHeaderMap = Utils.stream(segments1.get(0).getAnnotations().keySet().iterator()).filter(a -> !intersectingAnnotations.contains(a))
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Why do you need to use `Utils.stream` here and elsewhere?

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 3]
```diff
@@ -1104,9 +1111,199 @@ static public int beginOfShard(int shardIndex, int shardSize) {
...
+     *
+     * No copies of inputs are created
+     *
+     * @param keys -- the intervals we wish to query.  Sorted by interval.  No intervals overlap.  Never {@code null}
+     * @param vals -- the intervals that we wish to map to the keys.  Sorted by interval.  No intervals overlap.  Never {@code null}
+     * @return a mapping of intervals from keys to the list of overlapping intervals in vals.  All item in keys will
+     * have a key.  Never {@code null}
+     */
+    public static <T extends Locatable, U extends Locatable> Map<T, List<U>> createOverlapMap(final List<T> keys, final List<U> vals,
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: For small segment files, I'm more inclined to just rely on `OverlapDetector` to do this sort of thing.  Can you get rid of this method by just making an `OverlapDetector` for the unioned intervals in the calling code?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            Utils.stream(intersectingAnnotations.iterator()).forEach(a -> input1ToOutputHeaderMap.put(a, a + "_" + segmentsFileLabels.get(0)));
+
+            final Map<String, String> input2ToOutputHeaderMap = Utils.stream(segments2.get(0).getAnnotations().keySet().iterator()).filter(a -> !intersectingAnnotations.contains(a))
+                    .collect(Collectors.toMap(Function.identity(), Function.identity()));
+            Utils.stream(intersectingAnnotations.iterator()).forEach(a -> input2ToOutputHeaderMap.put(a, a + "_" + segmentsFileLabels.get(1)));
+
+            // Use the best available sequence dictionary, but if that is not available (i.e. null),
+            //   create one from the segment files.
+
+            final SAMSequenceDictionary sequenceDictionary = getBestAvailableSequenceDictionary() != null ?
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Actually, the sequence dictionary is never even used in `IntervalUtils.createOverlapMap`.

**Comment by samuelklee** [comment]
Actually, on second thought, even "union breakpoints" is misleading.  The result may have segment starts and ends that are not present in either input list, so it's not really a union.  Perhaps just use "combine" rather than "union" everywhere?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@CommandLineProgramProperties(
+        oneLineSummary = "(EXPERIMENTAL) Do a breakpoint union of two segment files and annotate with chosen columns from each file.",
+        summary = "Breakpoint union of two segment files while preserving annotations.\n" +
+                "This tool will load all segments into RAM.",
+        programGroup = CopyNumberProgramGroup.class)
+@BetaFeature
+public class UnionSegments extends GATKTool {
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Unit test as well.  Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+    static final String COLUMNS_OF_INTEREST_LONG_NAME = "columnsOfInterest";
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    static final String COLUMNS_OF_INTEREST_SHORT_NAME = "cols";
+    static final String LABELS_LONG_NAME = "labels";
+    static final String LABELS_SHORT_NAME = "lbls";
+    @Argument(
+            doc = "Input segment files -- must be specified twice, but order does not matter.",
+            fullName = ExomeStandardArgumentDefinitions.SEGMENT_FILE_LONG_NAME,
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done.  That is a reasonable rule.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            shortName = ExomeStandardArgumentDefinitions.SEGMENT_FILE_SHORT_NAME,
+            maxElements = 2, minElements = 2
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    )
+    protected List<File> segmentsFile;
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    @Argument(
+            doc = "Input segment file labels -- these will appear as suffixes in case of collisions.  The specification order must correspond to the input segment files.",
+            fullName = LABELS_LONG_NAME,
+            shortName = LABELS_SHORT_NAME,
+            maxElements = 2, minElements = 2, optional = true
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    )
+    protected List<String> segmentsFileLabels = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+    @Argument(
+            doc="List of columns in either segment file that should be reported in the output file.  If the column header exists in both, it will have an append.",
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+            fullName = COLUMNS_OF_INTEREST_LONG_NAME, shortName = COLUMNS_OF_INTEREST_SHORT_NAME, minElements = 1
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+    )
+    protected Set<String> columnsOfInterest = new HashSet<>();
+
+    @Argument(
+            doc="Output tsv file with union'ed segments",
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME)
+    protected File outputFile;
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: build.gradle [hunk 16]
```diff
@@ -144,6 +144,7 @@ dependencies {
     compile 'org.broadinstitute:barclay:1.2.1'
     compile 'com.github.broadinstitute:picard:' + picardVersion
     externalSourceConfiguration 'com.github.broadinstitute:picard:' + picardVersion + ':sources'
+    compile 'net.sf.supercsv:super-csv:2.4.0'
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply build.gradle]: No problem.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Not done...

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+    public void traverse() {
+        try {
+
+            final VersatileAnnotatedRegionParser parser = new VersatileAnnotatedRegionParser();
+            final List<SimpleAnnotatedGenomicRegion> segments1 = parser.readAnnotatedRegions(segmentsFile.get(0), columnsOfInterest);
+            final List<SimpleAnnotatedGenomicRegion> segments2 = parser.readAnnotatedRegions(segmentsFile.get(1), columnsOfInterest);
+            final List<List<SimpleAnnotatedGenomicRegion>> inputSegmentsLists = Arrays.asList(segments1,segments2);
+
+            if (segmentsFileLabels.size() == 0) {
+                segmentsFileLabels = IntStream.range(0, inputSegmentsLists.size()).mapToObj(i -> String.valueOf(i+1)).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+            }
+
+            // Check to see if we should warn the user that one or more columns of interest were not seen in any input file.
+            final Set<String> allSeenAnnotations = Sets.union(segments1.get(0).getAnnotations().keySet(), segments2.get(0).getAnnotations().keySet());
+            final Set<String> unusedColumnsOfInterest = Sets.difference(columnsOfInterest, allSeenAnnotations);
+            if (unusedColumnsOfInterest.size() > 0) {
+                final List<String> missingColumns = new ArrayList<>(unusedColumnsOfInterest);
+                logger.warn("Some columns of interest specified by the user were not seen in any input files: " + StringUtils.join(missingColumns, ", "));
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Now an exception is thrown.  
Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java (hunk 2 above)
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: You can't stream a set.  I'm open to suggestions.

No action.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Removed.  

Done.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+            final List<SimpleAnnotatedGenomicRegion> finalList = annotateUnionedIntervals(segments1, segments2,
+                    sequenceDictionary, Arrays.asList(input1ToOutputHeaderMap, input2ToOutputHeaderMap));
+
+            // TODO: Capture sample names in the comments if possible.
+            // TODO:  Allow choice in output names of interval headers.
+            final VersatileAnnotatedRegionParser writer = new VersatileAnnotatedRegionParser();
+            writer.writeAnnotatedRegionsAsTsv(finalList, outputFile, Collections.emptyList(),
+                    "contig", "start", "end");
+
+        } catch (final IOException ioe) {
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Whoops!  Fixed.

Done.  Did same for output.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java (hunk 2 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: You definitely can, but there are no guarantees on order unless the implementation guarantees it (e.g., LinkedHashSet/LinkedHashMap).  I suggest using such an implementation to guarantee the column order specified in the command line (as opposed to that in the files).  Add the corresponding test for good measure.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Did something equivalent.

Done.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+     *
+     * @return a list of simple annotated regions that is a union of segments1 and segments2 and contains all the annotations specified
+     *  in inputToOutputHeaderMaps.
+     */
+    private List<SimpleAnnotatedGenomicRegion> annotateUnionedIntervals(final List<SimpleAnnotatedGenomicRegion> segments1, final List<SimpleAnnotatedGenomicRegion> segments2,
+                                                                        final SAMSequenceDictionary dictionary, final List<Map<String, String>> inputToOutputHeaderMaps) {
+
+        final List<List<SimpleAnnotatedGenomicRegion>> segmentLists = Arrays.asList(segments1, segments2);
+
+        // We assume that the union'ed intervals are sorted.
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Added two new things:

- a new method for sorting the intervals in genomic order before doing the union
- a new comparator for this sorting.  Please note that it requires a SAMSequenceDictionary.

[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+        // TODO: This needs to be sorted if we want to support more than two sets of regions at once.
+        final List<Locatable> unionIntervals = IntervalUtils.unionIntervals(
+                segmentLists.get(0).stream().map(SimpleAnnotatedGenomicRegion::getInterval)
+                        .collect(Collectors.toList()),
+                segmentLists.get(1).stream().map(SimpleAnnotatedGenomicRegion::getInterval)
+                        .collect(Collectors.toList()));
+
+        // Create a list of maps where each entry corresponds to union'ed intervals to the regions in segmentList_i
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/UnionSegments.java]: Done

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 22]
```diff
@@ -53,7 +58,22 @@
...
     public static final Comparator<Locatable> LEXICOGRAPHICAL_ORDER_COMPARATOR =
-            Comparator.comparing(Locatable::getContig,Comparator.nullsLast(String::compareTo))
+            Comparator.comparing(Locatable::getContig, Comparator.nullsLast(String::compareTo))
+                    .thenComparingInt(Locatable::getStart)
+                    .thenComparingInt(Locatable::getEnd);
+
+    /**
+     * Please see:  https://github.com/gpanther/java-nat-sort
+     *
+     * Also called alpha numerical sorting.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: Can you not use the order imposed by an available sequence dictionary and the existing `IntervalUtils.compareLocatables(first, second, dictionary)` method?

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 23]
```diff
@@ -147,22 +167,23 @@ public static int compareContigs(final Locatable first, final Locatable second,
...
     /**
      * getSpanningInterval returns interval that covers all of the locations passed in.
+     *
      * @param locations the locations to be spanned (on a single contig)
      * @return the minimal span that covers all locations (could be null if no locations are passed in).
      * @throws IllegalArgumentException if the argument is null
-     * or if the argument contains any null element
-     * or if the locations are not all on the same contig (compared by String.equals)
+     *                                  or if the argument contains any null element
+     *                                  or if the locations are not all on the same contig (compared by String.equals)
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: Your IDE appears to have reformated this entire class -- this is bad because we want to preserve accurate provenance information for our code. Can you fix?

**Comment by LeeTL1220** [comment]
@samuelklee Removed references to union and replaced with combine
@droazen Undid the IDE formatting changes.
@droazen Refactored to use `IntervalUtils.compareLocatables(....)`

**Comment by LeeTL1220** [comment]
@samuelklee Back to you... finally.

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 24]
```diff
@@ -1109,4 +1113,262 @@ static public int endOfShard(int shardIndex, int shardSize) {
...
+
+    /**
+     *
+     * The order of contigs/sequences in the dictionary is the order of the sorting here.
+     *
+     * @param dictionary dictionary to use for the sorting.  Intervals with sequences not in this dictionary will cause
+     *                   exceptions to be thrown.
+     * @return an instance of {@code Comapator<Locatable>} for use in sorting of Locatables.
+     */
+    public static Comparator<Locatable> getDictionaryOrderComparator(final SAMSequenceDictionary dictionary) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: Non-null check.

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 25]
```diff
@@ -1109,4 +1113,262 @@ static public int endOfShard(int shardIndex, int shardSize) {
...
+            for (int i = 0; i < (breakpoints.size() - 1); i++) {
+                final int currentBreakpoint = breakpoints.get(i).getLeft();
+                final int nextBreakpoint = breakpoints.get(i + 1).getLeft();
+
+                final boolean isCurrentBreakpointStart = breakpoints.get(i).getRight() == IntervalBreakpointTypeEnum.START_BREAKPOINT;
+                final boolean isNextBreakpointStart = breakpoints.get(i + 1).getRight() == IntervalBreakpointTypeEnum.START_BREAKPOINT;
+
+                // if both breakpoints are starts of intervals, then the result is bp1, bp2-1
+                // if both breakpoints are ends of intervals, then the result is bp1+1, bp2
+                int start = (!isCurrentBreakpointStart && !isNextBreakpointStart ? currentBreakpoint + 1: currentBreakpoint);
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: White space before `:`.

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 26]
```diff
@@ -1109,4 +1113,262 @@ static public int endOfShard(int shardIndex, int shardSize) {
...
+                if (firstComparison != 0) {
+                    return firstComparison;
+                } else {
+                    // We want start before end
+                    return p1.getRight().compareTo(p2.getRight());
+                }
+            });
+            int numCurrentStarts = 0;
+            int numCurrentEnds = 0;
+            for (int i = 0; i < (breakpoints.size() - 1); i++) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: Just for the record, I have not gone through the logic here carefully.  This kind of code is pretty tricky (at least for me), and the less of it we have in the codebase the better, I think.  (I'm not even sure that my implementation in SegmentUtils was 100% bug free!)  I'm going to assume that the tests have covered all possible edge cases and that this code is completely correct for now...

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedRegionParser.java [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+     *  Write a tsv file of annotated regions with the specified strings for the position headers.
+     *
+     * @param regions list of locatables and associated annotations for writing.  One per line.
+     * @param outputFile location of the output file.
+     * @param comments list of comments to prepend to the tsv.  Each line will be prepended with "#"
+     */
+    public static void writeAnnotatedRegionsAsTsv(final List<SimpleAnnotatedGenomicRegion> regions, final File outputFile,
+                                           final List<String> comments) {
+
+        final List<String> columnHeaders = Lists.newArrayList(CONTIG_HEADER, START_HEADER, END_HEADER);
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedRegionParser.java]: Arrays.asList

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedRegionParser.java [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+import org.testng.collections.Lists;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+
+public class AnnotatedRegionParser {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedRegionParser.java]: Even if we are going to eventually write a Feature-based tool, I think we should make this a LocatableCollection when that PR is in.  I don't like having custom static methods for reading and writing each type of "Feature".  Could you file an issue?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+public class CombineSegmentBreakpoints extends GATKTool {
+
+    public static final String COLUMNS_OF_INTEREST_LONG_NAME = "columnsOfInterest";
+    public static final String COLUMNS_OF_INTEREST_SHORT_NAME = "cols";
+    public static final String LABELS_LONG_NAME = "labels";
+    public static final String LABELS_SHORT_NAME = "lbls";
+
+    @Argument(
+            doc = "Input segment files -- must be specified twice, but order does not matter.",
+            fullName = ExomeStandardArgumentDefinitions.SEGMENT_FILE_LONG_NAME,
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java]: Somewhere in the docs you should specify that `CONTIG`, `START`, and `END` are expected.  Probably here and in the class javadoc would be good.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java [hunk 30]
```diff
@@ -0,0 +1,165 @@
...
+    private List<File> segmentFiles = new ArrayList<>();
+    @Argument(
+            doc = "Input segment file labels -- these will appear as suffixes in case of collisions.  The specification order must correspond to the input segment files.",
+            fullName = LABELS_LONG_NAME,
+            shortName = LABELS_SHORT_NAME,
+            maxElements = 2,
+            minElements = 2,
+            optional = true
+    )
+    private List<String> segmentFileLabels = Lists.newArrayList("1", "2");
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java]: I think using `Arrays.asList` is more conventional in our codebase.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java [hunk 31]
```diff
@@ -0,0 +1,165 @@
...
+
+        final Map<String, String> input2ToOutputHeaderMap = segments2.get(0).getAnnotations().keySet().stream().filter(a -> !intersectingAnnotations.contains(a))
+                .collect(Collectors.toMap(Function.identity(), Function.identity()));
+        Utils.stream(intersectingAnnotations.iterator()).forEach(a -> input2ToOutputHeaderMap.put(a, a + "_" + segmentFileLabels.get(1)));
+
+        final List<SimpleAnnotatedGenomicRegion> finalList = annotateCombinedIntervals(segments1, segments2,
+                Arrays.asList(input1ToOutputHeaderMap, input2ToOutputHeaderMap), getBestAvailableSequenceDictionary());
+
+        // TODO: Capture sample names in the comments if possible.
+        AnnotatedRegionParser.writeAnnotatedRegionsAsTsv(finalList, outputFile, Collections.emptyList());
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java]: I'd prefer that you just add this capability, or remove the passing of comment strings for now.

File: src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java [hunk 32]
```diff
@@ -33,6 +36,7 @@
 public final class IntervalUtilsUnitTest extends BaseTest {
     public static final String INTERVAL_TEST_DATA = publicTestDir + "org/broadinstitute/hellbender/utils/interval/";
     public static final String emptyIntervals = BaseTest.publicTestDir + "empty_intervals.list";
+    public static final String FULL_HG19_DICT = BaseTest.publicTestDir + "Homo_sapiens_assembly19.dict";
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java]: Not sue why BaseTest needs to be specified in this line and in the line above?

File: src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java [hunk 33]
```diff
@@ -1386,4 +1390,409 @@ public void testToGatkIntervalString(){
             final SimpleInterval interval = new SimpleInterval("1",1,100);
             Assert.assertEquals(IntervalUtils.locatableToString(interval), "1:1-100");
     }
+
+    @DataProvider(name = "unionIntervalsTesting")
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java]: Same comment about "union" here and below.

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpointsIntegrationTest.java [hunk 34]
```diff
@@ -0,0 +1,189 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+public class CombineSegmentBreakpointsIntegrationTest extends CommandLineProgramTest {
+    private static final File TEST_RESOURCE_DIR = new File("src/test/resources/org/broadinstitute/hellbender/tools/copynumber/utils/");
+    public static final String SEG1 = TEST_RESOURCE_DIR.getAbsolutePath() + "/seg1.tsv";
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpointsIntegrationTest.java]: Note that I also want these test files to have "legacy" in the filename to indicate that the new workflows will not produce or consume files with those column headers.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedRegionParser.java (hunk 27 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedRegionParser.java]: There are a couple more instances elsewhere as well.

**Comment by LeeTL1220** [comment]
@samuelklee Back to you...

**Comment by LeeTL1220** [comment]
@samuelklee I think this is good for next review.

File: src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java [hunk 35]
```diff
@@ -1109,4 +1113,267 @@ static public int endOfShard(int shardIndex, int shardSize) {
...
+     */
+    public static Comparator<Locatable> getDictionaryOrderComparator(final SAMSequenceDictionary dictionary) {
+        Utils.nonNull(dictionary);
+        return (o1, o2) -> IntervalUtils.compareLocatables(o1, o2, dictionary);
+    }
+
+    /**
+     * An enum to classify breakpoints whether the breakpoint is the start or end of a region.
+     */
+    public enum IntervalBreakpointTypeEnum {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java]: IntervalBreakpointTypeEnum -> IntervalBreakpointType if you like.

File: src/test/resources/org/broadinstitute/hellbender/tools/copynumber/utils/combine-segment-breakpoints-with-legacy-header-learning-combined-copy-number.tsv [hunk 36]
```diff
@@ -0,0 +1,519 @@
..._36	learning_SAMPLE_37	learning_SAMPLE_38	learning_SAMPLE_39	learning_SAMPLE_40	learning_SAMPLE_41	learning_SAMPLE_42	learning_SAMPLE_43	learning_SAMPLE_44	learning_SAMPLE_45	learning_SAMPLE_46	learning_SAMPLE_47	learning_SAMPLE_48	learning_SAMPLE_49	learning_SAMPLE_50	learning_SAMPLE_51	learning_SAMPLE_52	learning_SAMPLE_53	learning_SAMPLE_54	learning_SAMPLE_55	learning_SAMPLE_56	learning_SAMPLE_57	learning_SAMPLE_58	learning_SAMPLE_59	learning_SAMPLE_60	learning_SAMPLE_61	learning_SAMPLE_62	learning_SAMPLE_63	learning_SAMPLE_64	learning_SAMPLE_65	learning_SAMPLE_66	learning_SAMPLE_67	learning_SAMPLE_68	learning_SAMPLE_69	learning_SAMPLE_70	learning_SAMPLE_71	learning_SAMPLE_72	learning_SAMPLE_73	learning_SAMPLE_74	learning_SAMPLE_75	learning_SAMPLE_76	learning_SAMPLE_77	learning_SAMPLE_78	learning_SAMPLE_79	learning_SAMPLE_80	learning_SAMPLE_81	learning_SAMPLE_82	learning_SAMPLE_83	learning_SAMPLE_84	learning_SAMPLE_85	learning_SAMPLE_86	learning_SAMPLE_87	learning_SAMPLE_88	learning_SAMPLE_89	learning_SAMPLE_90	learning_SAMPLE_91	learning_SAMPLE_92	learning_SAMPLE_93	learning_SAMPLE_94	learning_SAMPLE_95	learning_SAMPLE_96	learning_SAMPLE_97	learning_SAMPLE_98	learning_SAMPLE_99
```
Comment (samuelklee) [inline src/test/resources/org/broadinstitute/hellbender/tools/copynumber/utils/combine-segment-breakpoints-with-legacy-header-learning-combined-copy-number.tsv]: This test file seems a little on the big side, but I'll leave it up to you if you want to trim it down.

**Comment by samuelklee** [comment]
Just a couple of minor comments, looks good to go otherwise!  Thanks again for putting this together.

**Comment by LeeTL1220** [comment]
@samuelklee I have addressed both minor comments.  I am going to merge if tests pass.
